Configure object store for datasets in job handler instead of web thread.

jmchilton opened the pull request:
This is good for a few reasons:

* It speeds up the bottleneck that is the job creation code in the web thread (the job handler is much easier to parallelize). Skipping the flush for all those datasets and skipping the disk write per dataset should really speed things up nicely when submitting large numbers of jobs and jobs with a large number of inputs. Fixes https://github.com/galaxyproject/galaxy/issues/6513 and important progress on #4959.
* This prevents a disk slow down from grinding the web response to a halt. We don't want the handler to freeze up either - but if something is going to freeze it is a better user experience and easier to recover from the job handler freezing while their are disk problems.
* This allows per-job-handler object store configurations - simply configure different job handlers with different object store configurations. This is also a huge step toward per-job-destination object stores (see #6552 that builds on this PR to add that feature).

Updated: Rebased with some fixes and removed the refactoring toward per-job-destination commit since it has been merged into a single commit that actually adds that feature as part of #6552.

jmchilton committed:
Better encapsulation when queueing jobs.

No need for the runner to know all these internals of the JobWrapper when it is just taking attributes on the job or job wrapper and calling job wrapper methods with them. Provides nice symmetry between queue/finish now also.

jmchilton committed:
Setup objectstore for datasets in job handler instead of web thread.

This is good for a few reasons

* It speeds up the bottleneck that is the job creation code in the web thread (the job handler is much easier to parallelize). Skipping the flush for all those datasets and skipping the disk write per dataset should really speed things up nicely when submitting large numbers of jobs and jobs with a large number of inputs.
* This prevents a disk slow down from grinding the web response to a halt. We don't want the handler to freeze up either - but if something is going to freeze it is a better user experience and easier to recover from the job handler freezing while their are disk problems.
* This allows per-job-handler object store configurations - simply configure different job handlers with different object store configurations. This is cool and important and I can imagine some awesome applications but I will wait for them to materialize.

mvdbeek commented:
That is awesome, I also have the vague feeling that this could allow some cool extensions. I suppose a dynamic rule could be implemented that routes jobs to user or tool specific object stores? That would be really cool!

bgruening commented:
> This is cool and important and I can imagine some awesome applications but I will wait for them to materialize.

Indeed! Super cool, don't wait :) ... xref: https://github.com/galaxyproject/galaxy/issues/4974

jmchilton commented:
The new test case in #6546 is already paying dividends - I didn't write it for this PR but I included it here and it caught a very subtle problem with this PR that none of the other tests would have.

jmchilton committed:
Restructure JobWrapper/TaskWrapper for more unified initialization...

... especially after changes to set object store id in the job handler. This fixes the parallelism tests for those changes.

dannon committed:
Config rebuild, and a few typos.

VJalili commented on lib/galaxy/objectstore/__init__.py:
```diff
@@ -808,3 +808,24 @@ def _create_object_in_session(obj):
+            self.object_store.create(data.dataset)
+        except ObjectInvalid:
+            raise Exception('Unable to create output dataset: object store is full')
+        self.object_store_id = data.dataset.object_store_id  # these will be the same thing after the first output
```
@jmchilton so you expect the `data.dataset.object_store_id` to change after calling the `create` method?
